### PR TITLE
healthchecks: Create superuser idempotently

### PR DESCRIPTION
**Pull request by jdheyburn**
###### Description of changes

- create a superuser if the environment variables SUPERUSER_EMAIL and SUPERUSER_PASSWORD_FILE are set
- read from environment variable EMAIL_HOST_PASSWORD_FILE and set it to EMAIL_HOST_PASSWORD
  - this prevents the secret appearing in /nix/store

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/web-apps/healthchecks/default.nix [hunk 1]
```diff
@@ -57,13 +57,40 @@ py.pkgs.buildPythonApplication rec {
...
+    from django.contrib.auth.models import User;
+    from hc.accounts.views import _make_user;
+    import os
+    SUPERUSER_EMAIL = os.getenv("SUPERUSER_EMAIL")
+    SUPERUSER_PASSWORD_FILE = os.getenv("SUPERUSER_PASSWORD_FILE")
+    if SUPERUSER_PASSWORD_FILE:
+        with open(SUPERUSER_PASSWORD_FILE, "r") as file:
+            SUPERUSER_PASSWORD = file.readline()
+    if SUPERUSER_EMAIL and SUPERUSER_PASSWORD:
+        if User.objects.filter(email=SUPERUSER_EMAIL).count()==0:
```
Comment (phaer) [inline pkgs/servers/web-apps/healthchecks/default.nix:77]: spaces around `missing`

**Review by phaer — COMMENTED** [review]
LGTM in general, and I think it's nice to be able to declare a superuser. The thing is, i don't think we do that for (m?)any other django apps in nixpkgs. But as I am just the maintainer of this service and can't merge it myself anyway so I leave this to a nixpkgs committer to decide ;)

File: pkgs/servers/web-apps/healthchecks/default.nix (hunk 1 above)
Comment (jdheyburn) [reply pkgs/servers/web-apps/healthchecks/default.nix:77]: Do you mean to add spaces around `==`, like so?

```python
if User.objects.filter(email=SUPERUSER_EMAIL).count() == 0:
```

**Comment by jdheyburn** [comment]
@phaer Thanks - this is my first Django app I've deployed on NixOS so I'm unsure what the convention is.

The motivation for this change was a means of logging in to the service without having to perform any manual tasks. I had started with making EMAIL_HOST_PASSWORD secure and then having REGISTRATION_OPEN enabled, so then have a sign-up email sent to myself. That user wouldn't be a superuser however, so I started looking into how we can get a superuser created on start. I'm not sure of how other Django apps solve this problem.

File: pkgs/servers/web-apps/healthchecks/default.nix (hunk 1 above)
Comment (phaer) [reply pkgs/servers/web-apps/healthchecks/default.nix:77]: Yes

Comment (jdheyburn) [reply pkgs/servers/web-apps/healthchecks/default.nix]: Done 👍🏻

**Comment by jdheyburn** [comment]
@phaer Please let me know if there's anything else required to get this PR approved, so that it can be released into the wild :)

**Review by SuperSandro2000 — COMMENTED** [review]
Why are we not using EnvironmentFile?

**Comment by jdheyburn** [comment]
@SuperSandro2000 Thanks for the review!

I wasn't aware of `healthchecks-manage createsuperuser` but it looks like that does what I want already.

Since this would need `DJANGO_SUPERUSER_PASSWORD` to be set as an env var, it would make sense that this project is refactored to use an EnvironmentFile that the user provides, otherwise the secret would be in plaintext. Do other Django apps handle this problem like this?

I can have a stab at that in a separate PR, if that's something @phaer agrees with. Although it would be a breaking change.

**Comment by SuperSandro2000** [comment]
> Since this would need `DJANGO_SUPERUSER_PASSWORD` to be set as an env var, it would make sense that this project is refactored to use an EnvironmentFile that the user provides, otherwise the secret would be in plaintext. Do other Django apps handle this problem like this?

Environment and EnvironmentFile aren ot exclusive and can both be used at the same time.



> I can have a stab at that in a separate PR, if that's something @phaer agrees with. Although it would be a breaking change.

Feel free to create one. I can review and merge it.

**Comment by phaer** [comment]
Hey @jdheyburn Are you still interested in this? :)

**Comment by jdheyburn** [comment]
Hey @phaer thanks for the follow up - I am interested in it but I think it might be better to opt for @SuperSandro2000's approach to use the native env vars. I'll work on that in a separate PR. Feel free to close this PR out if you feel that is the better option :)

**Pull request closed without merging**